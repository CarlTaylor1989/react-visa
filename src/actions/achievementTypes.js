export const SET_COMPLETED_ACHIEVEMENT = 'SET_COMPLETED_ACHIEVEMENT';
export const RESTORE_ACHIEVEMENTS = 'RESTORE_ACHIEVEMENTS';

export const UPDATE_PRODUCT_ACHIEVEMENTS = 'UPDATE_PRODUCT_ACHIEVEMENTS';
export const UPDATE_PROPOSAL_ACHIEVEMENTS = 'UPDATE_PROPOSAL_ACHIEVEMENTS';
export const UPDATE_SCORED_ACHIEVEMENTS = 'UPDATE_SCORED_ACHIEVEMENTS';
export const UPDATE_SETTINGS_ACHIEVEMENTS = 'UPDATE_SETTINGS_ACHIEVEMENTS';
export const UPDATE_TIMED_ACHIEVEMENTS = 'UPDATE_TIMED_ACHIEVEMENTS';

export const RESET_CHANGED_ACHIEVEMENT = 'RESET_CHANGED_ACHIEVEMENT';
export const RESET_ACHIEVEMENTS = 'RESET_ACHIEVEMENTS';
