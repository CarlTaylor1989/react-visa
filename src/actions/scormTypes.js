export const INITIALISE_SCORM = 'INITIALISE_SCORM';

export const SET_LESSON_LOCATION = 'SET_LESSON_LOCATION';
export const SET_COMPLETION_STATUS = 'SET_COMPLETION_STATUS';
export const SET_SUSPEND_DATA_ITEM = 'GET_SUSPEND_DATA_ITEM';
export const SET_SCORM_VALUE = 'SET_SCORM_VALUE';

export const COMMIT = 'COMMIT';
export const TERMINATE_SCORM = 'TERMINATE_SCORM';

export const PREPARE_SUSPEND_DATA = 'PREPARE_SUSPEND_DATA';
