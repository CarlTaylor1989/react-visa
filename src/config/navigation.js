export const indexPath = '/';
export const splashPath = '/splash';
export const diagnosticPath = '/diagnostic';
export const introSequencePath = '/introseq';
export const gamePath = '/game';
export const networkPath = `${gamePath}/network`;
export const networkMapPath = `${networkPath}/map`;
export const networkProductPath = `${networkPath}/product`;
export const clientsPropsPath = `${gamePath}/clientproposals`;
export const clientsSelectorPath = `${clientsPropsPath}/selector`;
export const proposalPath = `${clientsPropsPath}/proposal`;

export const sequence = [
  {
    url: indexPath
  },
  {
    url: splashPath
  },
  {
    url: diagnosticPath
  },
  {
    url: introSequencePath
  },
  {
    url: gamePath,
    subpages: [{
      url: networkPath
    }, {
      url: clientsPropsPath
    }],
  }
];
