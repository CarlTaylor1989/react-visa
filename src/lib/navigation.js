import _ from 'lodash';
import {
  sequence,
  splashPath,
  diagnosticPath,
  introSequencePath,
  gamePath
} from '../config/navigation';

/**
 * Gets the next screen route
 * @param {string} current Current screen route
 * @returns {string}
 */
export const getNextScreen = (current) => {
  const currentIndex = _.findIndex(sequence, value => value.url === current);
  let nextPage = current;
  if (currentIndex > -1 && currentIndex < sequence.length - 1) {
    nextPage = sequence[currentIndex + 1].url;
  }
  return nextPage;
};

/**
 * Finds the next screen based on the intro and diagnostic completion
 * @param {boolean} diagnosticCompleted Diagnostic completion
 * @param {boolean} diagnosticVisited Diagnostic visited
 * @param {boolean} introCompleted Diagnostic visited
 * @param {string} current Current screen
 * @return {string}
 */
export const findNextScreen = (diagnosticCompleted, diagnosticVisited, introCompleted, current) => {
  let nextPage;
  if (diagnosticVisited && !diagnosticCompleted && current === splashPath) {
    nextPage = diagnosticPath;
  } else if (diagnosticCompleted && !introCompleted) {
    nextPage = introSequencePath;
  } else if (diagnosticCompleted && introCompleted) {
    nextPage = gamePath;
  } else {
    nextPage = splashPath;
  }
  return nextPage;
};
