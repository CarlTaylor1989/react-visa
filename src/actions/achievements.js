import * as types from './achievementTypes';

export function setCompletedAchievement(achievement) {
  return {
    type: types.SET_COMPLETED_ACHIEVEMENT,
    achievement
  };
}

export function restoreAchievements(completed, lastCompleted) {
  return {
    type: types.RESTORE_ACHIEVEMENTS,
    completed,
    lastCompleted
  };
}

export function updateProductAchievements() {
  return {
    type: types.UPDATE_PRODUCT_ACHIEVEMENTS
  };
}

export function updateProposalAchievements() {
  return {
    type: types.UPDATE_PROPOSAL_ACHIEVEMENTS
  };
}

export function updateScoredAchievements() {
  return {
    type: types.UPDATE_SCORED_ACHIEVEMENTS
  };
}

export function updateSettingsAchievements() {
  return {
    type: types.UPDATE_SETTINGS_ACHIEVEMENTS
  };
}

export function updateTimedAchievements() {
  return {
    type: types.UPDATE_TIMED_ACHIEVEMENTS
  };
}

export function resetChangedAchievement() {
  return {
    type: types.RESET_CHANGED_ACHIEVEMENT
  };
}

export function resetAchievements() {
  return {
    type: types.RESET_ACHIEVEMENTS
  };
}
