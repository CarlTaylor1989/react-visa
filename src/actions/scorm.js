import * as types from './scormTypes';

export function initialiseScorm() {
  return {
    type: types.INITIALISE_SCORM
  };
}

export function setLessonLocation(lessonLocation) {
  return {
    type: types.SET_LESSON_LOCATION,
    lessonLocation
  };
}

export function setCompletionStatus(completionStatus) {
  return {
    type: types.SET_COMPLETION_STATUS,
    completionStatus
  };
}

export function setSuspendDataItem(item, value) {
  return {
    type: types.SET_SUSPEND_DATA_ITEM,
    item,
    value
  };
}

export function setScormValue(item, value) {
  return {
    type: types.SET_SCORM_VALUE,
    item,
    value
  };
}

export function commit() {
  return {
    type: types.COMMIT
  };
}

export function terminateScorm() {
  return {
    type: types.TERMINATE_SCORM
  };
}

export function prepareSuspendData() {
  return {
    type: types.PREPARE_SUSPEND_DATA
  };
}
