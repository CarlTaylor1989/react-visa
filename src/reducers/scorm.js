import { createReducer } from 'redux-create-reducer';
import { SCORM, debug } from 'pipwerks-scorm-api-wrapper';
import { cloneDeep } from 'lodash';
import * as types from '../actions/scormTypes';
import config from '../tracking/scorm/config.json';
import * as initialState from '../tracking/scorm/scormInitialState.json';
import {
  getDataModel,
  validateCompletionStatus
} from '../tracking/scorm/DataModel';
import { indexPath } from '../config/navigation';

const commit = (suspendData, completionStatus, location) => {
  SCORM.set('cmi.suspend_data', JSON.stringify(suspendData));
  SCORM.set(getDataModel('location'), location);
  SCORM.status('set', completionStatus);
  SCORM.save();
};

const scormData = createReducer(cloneDeep(initialState), {
  [types.INITIALISE_SCORM](state) {
    let scormInitialised = state.scormInitialised;
    let learnerName = state.learnerName;
    let learnerID = state.learnerID;
    let lessonLocation = state.lessonLocation;
    let suspendData = state.suspendData;

    if (!state.scormInitialised) {
      debug.isActive = config.debugMode;
      const version = config.version.substr(0, 4) === '2004' ? '2004' : '1.2';
      SCORM.version = version;
      const scorm = SCORM.init();

      if (scorm) {
        scormInitialised = true;
        learnerName = SCORM.get(getDataModel('student_name'));
        learnerID = SCORM.get(getDataModel('student_id'));
        lessonLocation = SCORM.get(getDataModel('location'));
        lessonLocation = lessonLocation === '' ? indexPath : lessonLocation;
        suspendData = SCORM.get('cmi.suspend_data');
        suspendData = suspendData && suspendData.length > 0 ? JSON.parse(suspendData) : {};
      }
    }

    return {
      ...state,
      scormInitialised,
      learnerName,
      learnerID,
      lessonLocation,
      suspendData
    };
  },
  [types.SET_LESSON_LOCATION](state, action) {
    if (state.scormInitialised && state.lessonLocation !== action.lessonLocation) {
      SCORM.set(getDataModel('location'), action.lessonLocation);
    }
    return {
      ...state,
      lessonLocation: action.lessonLocation
    };
  },
  [types.SET_COMPLETION_STATUS](state, action) {
    return {
      ...state,
      completionStatus: validateCompletionStatus(state.completionStatus, action.completionStatus)
    };
  },
  [types.SET_SUSPEND_DATA_ITEM](state, action) {
    return {
      ...state,
      suspendData: {
        ...state.suspendData,
        [action.item]: action.value
      }
    };
  },
  [types.SET_SCORM_VALUE](state, action) {
    return {
      ...state,
      [action.item]: action.value
    };
  },
  [types.COMMIT](state) {
    if (state.scormInitialised) {
      const currentData = { ...state.suspendData };
      commit(currentData, state.completionStatus, state.lessonLocation);
    }

    return {
      ...state
    };
  },
  [types.TERMINATE_SCORM](state) {
    let scormInitialised = state.scormInitialised;

    if (state.scormInitialised) {
      const currentData = { ...state.suspendData };
      commit(currentData, state.completionStatus, state.lessonLocation);
      const success = SCORM.quit();
      if (success) {
        scormInitialised = false;
      }
    }

    return {
      ...state,
      scormInitialised
    };
  }
});

export default { scormData };
