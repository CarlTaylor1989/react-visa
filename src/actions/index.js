import { push } from 'connected-react-router';
import * as AchievementActions from './achievements';
import * as AudioActions from './audio';
import * as ClientActions from './client';
import * as DiagnosticActions from './diagnostic';
import * as GenericActions from './generic';
import * as HintsActions from './hints';
import * as MapActions from './map';
import * as PromptActions from './prompts';
import * as ScoreActions from './score';
import * as ScormActions from './scorm';
import * as SettingsActions from './settings';
import * as StreakActions from './streak';
import * as TutorialActions from './tutorial';

const ActionCreators = Object.assign(
  {},
  AchievementActions,
  AudioActions,
  ClientActions,
  DiagnosticActions,
  GenericActions,
  HintsActions,
  MapActions,
  PromptActions,
  ScoreActions,
  ScormActions,
  SettingsActions,
  StreakActions,
  TutorialActions,
  { push }
);

export default ActionCreators;
