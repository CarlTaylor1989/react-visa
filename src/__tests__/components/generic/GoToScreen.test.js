/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import {
  GoToScreen,
  mapStateToProps,
  mapDispatchToProps
} from '../../../components/generic/GoToScreen';
import * as actions from '../../../actions/scorm';
import * as types from '../../../actions/scormTypes';
import { getNextScreen, findNextScreen } from '../../../lib/navigation';

jest.mock('../../../lib/navigation');

let mockProps;
beforeEach(() => {
  jest.resetAllMocks();
  mockProps = {
    commit: jest.fn(),
    diagnosticCompleted: false,
    diagnosticVisited: false,
    introSeqCompleted: false,
    lessonLocation: '',
    prepareSuspendData: jest.fn(),
    push: jest.fn(),
    resetDiagnostic: jest.fn(),
    setLessonLocation: jest.fn(),
    setPlayAudio: jest.fn()
  };
});

test('renders the Splash component with a start button', () => {
  const wrapper = shallow(<GoToScreen {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('method onChangeRegion()', () => {
  // setup
  const wrapper = shallow(<GoToScreen {...mockProps} />);
  wrapper.instance().onGoToScreen = jest.fn();

  // function under test
  wrapper.instance().onChangeRegion();

  // expectations
  expect(mockProps.resetDiagnostic).toHaveBeenCalled();
  expect(wrapper.instance().onGoToScreen).toHaveBeenCalledWith(expect.any(String));
});

describe('method onGoToScreen()', () => {
  test('sets the specified screen to the lesson location and navigates to that screen', () => {
    const newScreenId = 'welcome';
    const wrapper = shallow(<GoToScreen {...mockProps} />);
    wrapper.instance().onGoToScreen(newScreenId);
    expect(mockProps.setLessonLocation).toHaveBeenCalledWith(newScreenId);
  });

  test('invalid screen id passed', () => {
    const wrapper = shallow(<GoToScreen {...mockProps} />);
    wrapper.instance().onGoToScreen();
    expect(mockProps.setLessonLocation).not.toHaveBeenCalledWith();
  });
});

test('method onHoverBtn()', () => {
  const wrapper = shallow(<GoToScreen {...mockProps} />);
  wrapper.instance().onHoverBtn();
  expect(mockProps.setPlayAudio).toHaveBeenCalledWith(expect.any(String));
});

test('method goToNextScreen()', () => {
  mockProps.lessonLocation = 'irrelevant';
  const wrapper = shallow(<GoToScreen {...mockProps} />);
  wrapper.instance().goToScreenAfter = jest.fn();
  wrapper.instance().goToNextScreen();
  expect(wrapper.instance().goToScreenAfter).toHaveBeenCalledWith(mockProps.lessonLocation);
});

test('method goToScreenAfter()', () => {
  // setup
  const screenId = 'foo';
  const nextScreen = 'bar';
  getNextScreen.mockReturnValue(nextScreen);
  const wrapper = shallow(<GoToScreen {...mockProps} />);
  wrapper.instance().onGoToScreen = jest.fn();

  // function under test
  wrapper.instance().goToScreenAfter(screenId);

  // expectations
  expect(getNextScreen).toHaveBeenCalledWith(screenId);
  expect(wrapper.instance().onGoToScreen).toHaveBeenCalledWith(nextScreen);
});

test('method findAndGoToScreen()', () => {
  // setup
  const nextScreen = 'irrelevant';
  findNextScreen.mockReturnValue(nextScreen);
  const wrapper = shallow(<GoToScreen {...mockProps} />);
  wrapper.instance().onGoToScreen = jest.fn();

  // function under test
  wrapper.instance().findAndGoToScreen();

  // expectations
  expect(findNextScreen).toHaveBeenCalledWith(
    expect.any(Boolean),
    expect.any(Boolean),
    expect.any(Boolean),
    expect.any(String)
  );
  expect(wrapper.instance().onGoToScreen).toHaveBeenCalledWith(nextScreen);
});

test('method goToSplashScreen()', () => {
  // setup
  const wrapper = shallow(<GoToScreen {...mockProps} />);
  wrapper.instance().onGoToScreen = jest.fn();

  // function under test
  wrapper.instance().goToSplashScreen();

  // expectation
  expect(wrapper.instance().onGoToScreen).toHaveBeenCalledWith(expect.any(String));
});

test('should mapStateToProps', () => {
  const state = {
    diagnosticData: {
      completed: false
    },
    scormData: {
      lessonLocation: ''
    },
    genericData: {
      screen: ''
    }
  };
  const props = mapStateToProps(state);
  expect(props).toEqual({
    ...state.scormData,
    ...state.genericData,
    diagnosticCompleted: state.diagnosticData.completed
  });
});

test('should mapDispatchToProps', () => {
  const dispatch = jest.fn();
  jest.spyOn(actions, 'setLessonLocation').mockReturnValue('setLessonLocation()');
  const props = mapDispatchToProps(dispatch);
  props.setLessonLocation();
  expect(dispatch).toHaveBeenCalledWith({
    type: types.SET_LESSON_LOCATION
  });
});
