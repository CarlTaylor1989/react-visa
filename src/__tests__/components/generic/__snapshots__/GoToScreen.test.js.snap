// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Splash component with a start button 1`] = `""`;
